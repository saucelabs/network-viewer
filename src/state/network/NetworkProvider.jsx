import React, { useEffect, useReducer, useMemo } from 'react';
import PropTypes from 'prop-types';

import { reducer, initialState as defaultState } from './reducer';
import { updateData, fetchFile, updateScrollToIndex } from './actions';
import { NetworkContext } from './Context';
import { findRequestIndex } from '../../utils';
import { ROW_ID_PREFIX } from '../../constants';

const NetworkProvider = (props) => {
  const {
    data,
    file,
    fetchOptions,
    initialState,
    scrollTimeStamp,
    scrollRequestPosition,
    autoHighlightChange,
    onDataLoaded,
    onDataError,
    onPause,
    onResume,
    onReset,
  } = props;

  const [state, dispatch] = useReducer(reducer, initialState);
  const callbacks = {
    onPause,
    onResume,
    onReset,
  };
  const value = useMemo(() => [state, dispatch, callbacks], [state]);
  const selectedReqIndex = value[0].get('selectedReqIndex');
  const requestData = value[0].get('data');
  const showReqDetail = value[0].get('showReqDetail');
  const actualData = value[0].get('actualData');
  const error = value[0].get('error');

  // Update data onChange of network data
  useEffect(() => {
    if (data && data.log && data.log.entries) {
      updateData(dispatch)(data);
    }
  }, [data]);

  // Fetch HAR file onChange of file prop
  useEffect(() => {
    if (file) {
      fetchFile(dispatch)(file, fetchOptions);
    }
  }, [file]);

  useEffect(() => {
    if (actualData.size && onDataLoaded) {
      onDataLoaded(actualData);
    }
  }, [actualData]);

  useEffect(() => {
    if (error && onDataError) {
      onDataError(error);
    }
  }, [error]);

  // Find nearby request-rowId and update scrollIndex on scrollTimeStamp receive
  useEffect(() => {
    const shouldChangeHighlight = showReqDetail ? autoHighlightChange : true;
    if (scrollTimeStamp && shouldChangeHighlight) {
      const reqIndex = findRequestIndex({
        data: requestData,
        timestamp: scrollTimeStamp,
        position: scrollRequestPosition,
      });
      if (reqIndex || reqIndex === 0) {
        updateScrollToIndex(dispatch)(requestData.get(reqIndex));
      }
    }
  }, [scrollTimeStamp, actualData]);

  // Scroll to request row onChange of scrollToIndex
  // setTimeout is required when reqDetail is visible,
  // it allows DOM to adjust before we scroll it to highlighted request
  useEffect(() => {
    if (selectedReqIndex) {
      setTimeout(() => {
        const element = document.getElementById(ROW_ID_PREFIX + selectedReqIndex);
        if (element) {
          element.scrollIntoView({
            alignToTop: true,
            behavior: 'smooth',
          });
        }
      }, 300);
    }
  }, [selectedReqIndex]);

  return (
    <NetworkContext.Provider
      value={value}
      {...props}
    />
  );
};

NetworkProvider.propTypes = {
  autoHighlightChange: PropTypes.bool,
  data: PropTypes.object,
  fetchOptions: PropTypes.object,
  file: PropTypes.string,
  initialState: PropTypes.object,
  onDataError: PropTypes.func,
  onDataLoaded: PropTypes.func,
  onPause: PropTypes.func,
  onReset: PropTypes.func,
  onResume: PropTypes.func,
  scrollRequestPosition: PropTypes.oneOf(['before', 'after', 'near']),
  scrollTimeStamp: PropTypes.number,
};

NetworkProvider.defaultProps = {
  autoHighlightChange: false,
  data: null,
  fetchOptions: { withCredentials: true },
  file: null,
  initialState: defaultState,
  onDataError: null,
  onDataLoaded: null,
  onPause: null,
  onReset: null,
  onResume: null,
  scrollRequestPosition: 'near',
  scrollTimeStamp: null,
};

export default NetworkProvider;
