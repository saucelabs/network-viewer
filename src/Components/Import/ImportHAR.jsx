import React from 'react';
import PropTypes from 'prop-types';
import { useDropzone } from 'react-dropzone';

import { useNetwork } from './../../state/network/Context';
import Styles from './ImportHAR.styles.scss';
import Button from './../Common/Button';

const DROP_FILE_CONFIG = {
  accept: '.har',
  multiple: false,
};

const ImportHar = ({ showButton, className }) => {
  const { actions } = useNetwork();
  const { updateErrorMessage } = actions;

  const prepareData = (newNetworkData) => (
    actions.updateData({
      entries: newNetworkData.log.entries,
      pages: newNetworkData.log.pages,
    })
  );

  const onDrop = (files) => {
    const reader = new FileReader();
    reader.onabort = () => updateErrorMessage({ title: 'file reading was aborted' });
    reader.onerror = () => updateErrorMessage({ title: 'file reading has failed' });
    reader.onload = () => {
      try {
        const data = JSON.parse(reader.result);
        prepareData(data);
      } catch (error) {
        updateErrorMessage({ title: 'Error while parsing HAR file' });
      }
    };
    reader.readAsText(files[0]);
  };

  const { getRootProps, getInputProps } = useDropzone({
    ...DROP_FILE_CONFIG,
    onDrop,
  });

  return (
    <div {...getRootProps()}>
      <input {...getInputProps()} />
      {showButton ? (
        <Button
          category="default"
          className={className}
          material
          raised
          size="sm"
        >
          Import HAR
        </Button>
      ) : (
        <p className={Styles['drag-drop']}>Drag and drop HAR file here, or click to select file</p>
      )}
    </div>
  );
};

ImportHar.propTypes = {
  className: PropTypes.string,
  showButton: PropTypes.bool,
};

ImportHar.defaultProps = {
  className: null,
  showButton: true,
};

export default ImportHar;
