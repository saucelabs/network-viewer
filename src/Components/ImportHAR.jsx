import React from 'react';
import PropTypes from 'prop-types';
import { useDropzone } from 'react-dropzone';
import { Button } from 'react-bootstrap';
import { useNetwork } from '../state/network/provider';

import Styles from './ImportHAR.module.scss';

const DROP_FILE_CONFIG = {
  accept: '.har',
  multiple: false,
};

const ImportHar = ({ showButton }) => {
  const { actions } = useNetwork();

  const prepareData = newNetworkData => (
    actions.updateData(newNetworkData.log.entries)
  );

  const onDrop = (files) => {
    const reader = new FileReader();
    reader.onabort = () => actions.errorNotification({ description: 'file reading was aborted' });
    reader.onerror = () => actions.errorNotification({ description: 'file reading has failed' });
    reader.onload = () => {
      try {
        const data = JSON.parse(reader.result);
        prepareData(data);
      } catch (error) {
        actions.errorNotification({ description: 'Error while parsing HAR file' });
      }
    };
    reader.readAsText(files[0]);
  };

  const { getRootProps, getInputProps } = useDropzone({
    ...DROP_FILE_CONFIG,
    onDrop,
  });

  return (
    <div {...getRootProps()}>
      <input {...getInputProps()} />
      {
        showButton
          ? <Button variant="secondary" size="sm">Import HAR file</Button>
          : <p className={Styles['drag-drop']}>Drag and drop HAR file here, or click to select file</p>
      }
    </div>
  );
};

ImportHar.propTypes = {
  showButton: PropTypes.bool,
};

ImportHar.defaultProps = {
  showButton: true,
};

export default ImportHar;
