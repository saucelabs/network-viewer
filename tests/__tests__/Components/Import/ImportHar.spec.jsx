import React from 'react';
import { mount } from 'enzyme';

import ImportHAR from './../../../../src/Components/Import/ImportHAR';
import NetworkProvider from './../../../../src/state/network/NetworkProvider';

describe('ImportHAR', () => {
  it('renders without crashing', () => {
    mount(
      <NetworkProvider>
        <ImportHAR />
      </NetworkProvider>,
    );
  });
});
