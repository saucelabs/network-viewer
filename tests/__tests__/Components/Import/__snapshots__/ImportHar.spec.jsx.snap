// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportHAR renders without crashing 1`] = `
<NetworkProvider
  autoHighlightChange={false}
  data={null}
  fetchOptions={
    Object {
      "withCredentials": true,
    }
  }
  file={null}
  initialState={
    Immutable.Map {
      "errorFilter": false,
      "search": Object {
        "name": "url",
        "value": "",
      },
      "error": null,
      "scrollToIndex": null,
      "data": Immutable.List [],
      "selectedReqIndex": null,
      "reqDetail": null,
      "showReqDetail": false,
      "totalNetworkTime": null,
      "actualData": Immutable.List [],
      "filter": Object {
        "name": null,
        "value": null,
      },
      "dataSummary": Immutable.Map {},
      "loading": false,
      "sort": Object {
        "isAcs": true,
        "key": "startedDateTime",
      },
    }
  }
  onDataError={null}
  onDataLoaded={null}
  scrollRequestPosition="near"
  scrollTimeStamp={null}
>
  <ImportHar
    className={null}
    showButton={true}
  >
    <div
      onBlur={[Function]}
      onClick={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <input
        accept=".har"
        autoComplete="off"
        multiple={false}
        onChange={[Function]}
        onClick={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        tabIndex={-1}
        type="file"
      />
      <Button
        className={null}
        disabled={false}
        href={null}
        type="button"
        variant="default"
      >
        <button
          className="btn"
          disabled={false}
          href={null}
          type="button"
        >
          Import HAR
        </button>
      </Button>
    </div>
  </ImportHar>
</NetworkProvider>
`;
